File: LeetCode/994.ts
function orangesRotting(grid: number[][]): number {
    const hasFresh = () => {
        for (let i = 0; i < grid.length; i++){
            for (let j = 0; j < grid[0].length; j++){
                if (grid[i][j] === 1){
                    return true;
                }
            }
        }
        return false;
    };
    const isFreshOrange = (row, col) => {
        if (row < 0 || row >= grid.length || col < 0 || col >= grid[0].length){
            return false;
        }

        return grid[row][col] === 1;
    };

    let currentCycle = [];
    for (let i = 0; i < grid.length; i++){
        for (let j = 0; j < grid[0].length; j++){
            if (grid[i][j] === 2){
                currentCycle.push([i, j]);
            }
        }
    }

    let steps = 0;
    while (currentCycle.length > 0){
        steps++;
        let nextCycle = [];
        for (let [row, col] of currentCycle){
            for (let [nextRow, nextCol] of [[row, col-1], [row, col+1], [row+1, col], [row-1, col]]){
                if (!isFreshOrange(nextRow, nextCol)){
                    continue;
                }

                grid[nextRow][nextCol] = 2;
                nextCycle.push([nextRow, nextCol]);
            }
        }
        currentCycle = nextCycle;
        // console.log("grid", grid);
        // console.log("nextCycle", nextCycle);
    }

    if (hasFresh()){
        return -1;
    }
    else {
        return Math.max(steps-1, 0);
    }
};

File: LeetCode/994.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        n = len(grid[0])
        m = len(grid)
        levels = 0
        DX = [-1, 1, 0, 0]
        DY = [0, 0, -1, 1]

        curr = []
        hasOrange = False
        for row in range(m):
            for col in range(n):
                if grid[row][col] >= 1:
                    hasOrange = True
                if grid[row][col] == 2:
                    curr.append((row, col))

        if not hasOrange:
            return 0

        while curr:
            # print("curr", curr)
            levels += 1
            later = []
            for orange in curr:
                for i in range(4):
                    nR = orange[0] + DX[i]
                    nC = orange[1] + DY[i]
                    if nR < 0 or nR >= m or nC < 0 or nC >= n:
                        continue
                    if grid[nR][nC] != 1:
                        continue
                    grid[nR][nC] = 2
                    later.append((nR, nC))
            curr = later

        for row in range(m):
            for col in range(n):
                if grid[row][col] == 1:
                    return -1

        return levels - 1